<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">

<!-- Begin Jekyll SEO tag v2.4.0 -->
<title>ukbtools | An R package to manipulate and explore UK Biobank data</title>
<meta name="generator" content="Jekyll v3.7.3" />
<meta property="og:title" content="ukbtools" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="An R package to manipulate and explore UK Biobank data" />
<meta property="og:description" content="An R package to manipulate and explore UK Biobank data" />
<link rel="canonical" href="http://localhost:4000/" />
<meta property="og:url" content="http://localhost:4000/" />
<meta property="og:site_name" content="ukbtools" />
<script type="application/ld+json">
{"url":"http://localhost:4000/","headline":"ukbtools","description":"An R package to manipulate and explore UK Biobank data","name":"ukbtools","@type":"WebSite","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->

    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#157878">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" href="/assets/css/style.css?v=0f70c681878c1fc5467ac36623e0eb843a0cbb36">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">ukbtools</h1>
      <h2 class="project-tagline">An R package to manipulate and explore UK Biobank data</h2>
      
        <a href="http://github.com/kenhanscombe/ukbtools" class="btn">View on GitHub</a>
      
      <a href="https://cran.r-project.org/web/packages/ukbtools/index.html" class="btn">View on CRAN</a>
      
    </section>

    <section class="main-content">
      
<p>The <strong>UK Biobank</strong> is a resource that includes detailed health-related and genetic data on about 500,000 individuals and is available to the research community. <strong>ukbtools</strong> removes all the upfront data wrangling required to get a single dataset for statistical analysis, and provides tools to assist in quality control, query of disease diagnoses, and retrieval of genetic metadata.</p>

<p><br /></p>

<h2 id="getting-started">Getting started</h2>

<p>Download and decrypt your data with the supplied <a href="https://biobank.ctsu.ox.ac.uk/crystal/exinfo.cgi?src=accessing_data_guide#convert">helper programs</a>. To use ukbtools, you need to create a UKB fileset (.tab, .r, and .html):</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_unpack ukbxxxx.enc key
ukb_conv ukbxxxx.enc_ukb r
ukb_conv ukbxxxx.enc_ukb docs
</code></pre></div></div>

<p>ukb_unpack decrypts your downloaded ukbxxxx.enc file, outputting a ukbxxxx.enc_ukb file. ukb_conv with the r flag converts the decrypted data to a tab-delimited file ukbxxxx.tab and an R script ukbxxxx.r that reads the tab file. The docs flag creates an html file containing a field-code-to-description table (among others).</p>

<p><strong>Note.</strong> Full details of the data download and decrypt process are given in the <a href="https://biobank.ctsu.ox.ac.uk/crystal/exinfo.cgi?src=accessing_data_guide">Using UK Biobank Data</a> documentation . <a href="https://biobank.ctsu.ox.ac.uk/crystal/download.cgi">Updated versions</a> of these helper programs exist. Other than small name changes (underscores removed) they appear to function similarly.</p>

<p><br /></p>

<h2 id="installing-the-package">Installing the package</h2>

<p>In R,</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Install from CRAN
install.packages("ukbtools")

# Install latest development version
devtools::install_github("kenhanscombe/ukbtools", build_vignettes = TRUE, dependencies = TRUE)
</code></pre></div></div>

<p><br /></p>

<h2 id="making-a-dataset">Making a dataset</h2>

<p>The function <code class="highlighter-rouge">ukb_df()</code> takes the stem of your fileset and returns a dataframe with usable column names.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(ukbtools)

my_ukb_data &lt;- ukb_df("ukbxxxx")
</code></pre></div></div>

<p>You can also specify the path to your fileset if it is not in the current directory. For example, if your fileset is in a subdirectory of the working directory called <code class="highlighter-rouge">data</code></p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_ukb_data &lt;- ukb_df("ukbxxxx", path = "/full/path/to/my/ukb/fileset/data")
</code></pre></div></div>

<p><br /></p>

<h2 id="making-a-key">Making a key</h2>

<p>Use <code class="highlighter-rouge">ukb_df_field</code> to create a field code-to-descriptive name key, as dataframe or named lookup vector.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_ukb_key &lt;- ukb_df_field("ukbxxxx", path = "/full/path/to/my/ukb/fileset/data")
</code></pre></div></div>

<p><strong>Note.</strong> You can move the three files in your fileset after creating them with <code class="highlighter-rouge">ukb_conv</code>, but they should be kept together. <code class="highlighter-rouge">ukb_df()</code> automatically updates the read call in the R source file to point to the correct directory (the current directly by default, or the directory specified by <code class="highlighter-rouge">path</code>).</p>

<p><br /></p>

<blockquote>
  <p><strong>Memory and efficiency</strong></p>

  <p>To reduce you memory usage, you could save your new UKB dataset with <code class="highlighter-rouge">save(my_ukb_data, file = "my_ukb_data.rda")</code>. Load the dataset with <code class="highlighter-rouge">load("my_ukb_data.rda")</code>. A UKB dataset from my largest UKB fileset which included a 2.6 GB .tab file took a little under 2 minutes to create with <code class="highlighter-rouge">ukb_df</code>. The associated .rda file was 138 MB and loaded in a little under 1.5 mins.</p>
</blockquote>

<p><br /></p>

<h2 id="multiple-downloads">Multiple downloads</h2>

<p>If you have multiple UKB downloads, first read each one in, then merge them with your preferred method. You could use <code class="highlighter-rouge">ukb_df_full_join</code> which is a thin wrapper around <code class="highlighter-rouge">dplyr::full_join</code> applied recursively with <code class="highlighter-rouge">purrr::reduce</code>.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukbxxxx_data &lt;- ukb_df("ukbxxxx")
ukbyyyy_data &lt;- ukb_df("ukbyyyy")
ukbzzzz_data &lt;- ukb_df("ukbzzzz")

ukb_df_full_join(ukbxxxx_data, ukbyyyy_data, ukbzzzz_data)
</code></pre></div></div>

<p><br /></p>

<blockquote>
  <p><strong>Repeated variables.</strong></p>

  <p>The join key is set to “eid” only (default value of the <code class="highlighter-rouge">by</code> parameter). Any additional variables common to any two tables will have “.x” and “.y” appended to their names. If you are satisfied the additional variables are identical to the original, the copies can be safely deleted. For example, if <code class="highlighter-rouge">setequal(my_ukb_data$var, my_ukb_data$var.x)</code> is <code class="highlighter-rouge">TRUE</code>, then my_ukb_data$var.x can be dropped. A <code class="highlighter-rouge">dlyr::full_join</code> is like the set operation union in that all abservation from all tables are included, i.e., all samples are included even if they are not included in all datasets.</p>

  <p>Repeated variable names <strong>within</strong> UKB datasets are unlikely to occur. <code class="highlighter-rouge">ukb_df</code> creates variable names by combining a snake_case descriptor with the variable’s <em>index</em> and <em>array</em>. This should be sufficient to uniquely identify the variable. However, if an <em>index_array</em> combination is incorrectly repeated in the original UKB data, this will result in a duplicated variable name. We observed two instances. The variables were encoded *<field>–0.0*, *<field>–1.0*, *<field>––1.0*, and `ukb_df` created a variable named *var_0_0*, *var_1_0*, *var_1_0*. This is probably a typo that should have been *<field>–0.0*, *<field>–1.0*, *<field>–2.0*, consistent with UKB official documentation describing the field as having 3 values for index. We have provided `ukb_df_duplicated_names` to identify duplicated names within a dataset. This will allow the user to make changes as appropriate. We expect the occurrence of such duplicates will be rare.</field></field></field></field></field></field></p>
</blockquote>

<p><br /></p>

<h2 id="exploring-primary-demographics-of-a-ukb-subset">Exploring primary demographics of a UKB subset</h2>

<p>As an exploratory step you might want to look at the demographics of a particular subset of the UKB sample relative to a reference sample. For example, using the <code class="highlighter-rouge">nonmiss.var</code> argument of <code class="highlighter-rouge">ukb_context</code> will produce a plot of the primary demographics (sex, age, ethnicity, and Townsend deprivation score) and employment status and assessment centre, for the subsample with data on your variable of interest compared to those without data (i.e. <code class="highlighter-rouge">NA</code>).</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_context(my_ukb_data, nonmiss.var = "my_variable_of_interest")
</code></pre></div></div>

<p align="center">
<img src="img/ukb_context_stack_111017.png" width="800px" />
</p>

<p><strong>Figure 1</strong> Primary demographic data for a UKB subset of interest.</p>

<p><br /></p>

<p>It is also possible to supply a logical vector with <code class="highlighter-rouge">subset.var</code> to define the subset and reference sample. This is particularly useful for understanding a subgroup within the UKB study, e.g., obese individuals below age 50.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>subgroup_of_interest &lt;- (my_ukb_data$bmi &gt; 40 &amp; my_ukb_data$age &lt; 50) 
ukb_context(my_ukb_data, subset.var = subgroup_of_interest)
</code></pre></div></div>

<p><br /></p>

<h2 id="retrieving-icd-diagnoses">Retrieving ICD diagnoses</h2>

<p>All ICD related functions begin <code class="highlighter-rouge">ukb_icd_</code>. Type <code class="highlighter-rouge">ukb_icd_</code> tab to see the family of functions. The full ICD “code-meaning” tables are available as datasets (<code class="highlighter-rouge">icd9codes</code>, <code class="highlighter-rouge">icd10codes</code>). ICD chapter-level tables describing disease blocks are also available for query as datasets (<code class="highlighter-rouge">icd9chapters</code>, <code class="highlighter-rouge">icd10chapters</code>)</p>

<p>To retrieve the full diagnosis of an individual (combine multiple individuals with <code class="highlighter-rouge">c()</code>)</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_icd_diagnosis(my_ukb_data, id = "0000000", icd.version = 10)
</code></pre></div></div>

<p>To retrieve the “meaning” of an ICD code use <code class="highlighter-rouge">icd_code</code>. Again, you can look up multiple codes by combining them with <code class="highlighter-rouge">c()</code>.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_icd_code_meaning(icd.code = "I74", icd.version = 10)
</code></pre></div></div>

<p>Search for a class of diseases with a keyword. Supplying multiple keywords with <code class="highlighter-rouge">c()</code>, will return all ICD entries containing <em>any</em> of the keywords.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_icd_keyword("cardio", icd.version = 10)
</code></pre></div></div>

<p>You can calculate the prevalence of a diagnosis in the UKB study (or a subset of the full sample) using <code class="highlighter-rouge">ukb_icd_prevalence</code>. The <code class="highlighter-rouge">icd.diagnosis</code> argument takes a regular expression, and so can also be used to retrieve prevalence of a disease “class”, i.e., the proportion of individuals with <em>any</em> diagnosis in the disease class.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code># ICD-10 code I74, Arterial embolism and thrombosis
ukb_icd_prevalence(my_ukb_data, icd.version = 10, icd.diagnosis = "I74")

# ICD-10 chapter 9, disease block I00–I99, Diseases of the circulatory system
ukb_icd_prevalence(my_ukb_data, icd.version = 10, icd.diagnosis = "I")

# ICD-10 chapter 2, C00-D49, Neoplasms 
ukb_icd_prevalence(my_ukb_data, icd.version = 10, icd.diagnosis = "C|D[0-4].")
</code></pre></div></div>

<p>To retrieve frequency for one or more ICD diagnoses by the levels of a reference variable, e.g., sex (male or female) use <code class="highlighter-rouge">ukb_icd_freq_by</code>. If the variable is continuous, it is divided into N approximately equal-sized groups (default = 10) within which ICD diagnosis frequency is calculated. ukb_icd_freq_by also includes an option to produce a figure of ICD diagnosis frequency by reference variable. Diagnoses of interest are passed to <code class="highlighter-rouge">icd.code</code>. The default ICD codes are the WHO top 3 cause of death worldwide (2015): coronary artery disease (CAD), cerebrovascular disease/ stroke, lower respiratory tract infection (LTRI).</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code># To plot the frequency of the default ICD codes with respect to BMI
ukb_icd_freq_by(my_ukb_data, reference.var = "body_mass_index_bmi_0_0", freq.plot = TRUE)
</code></pre></div></div>

<p>Setting <code class="highlighter-rouge">freq.plot = FALSE</code> (default) returns a dataframe of the frequencies. Values for the reference variable group ranges are in the column “group”.</p>

<p><br /></p>

<h2 id="retrieving-genetic-metadata">Retrieving genetic metadata</h2>

<p>If you are doing any downstream genetic analyses, you will need the genetic metadata (which should be in you phenotype dataset). Detailed information is available on UKB <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/UKBiobank_genotyping_QC_documentation-web.pdf">genotyping and quality control</a> and <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/imputation_documentation_May2015.pdf">imputation and association</a>.</p>

<p>All genetic metadata related functions begin <code class="highlighter-rouge">ukb_gen_</code>. Typing <code class="highlighter-rouge">ukb_gen_</code> tab opens a dropdown menu for this group of functions.</p>

<p><br /></p>

<p>You can collect the genetic metadata (including recommended exclusions, genetic sex, genetic ethnicity, chip, etc.), and principal components with</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_gen_meta &lt;- ukb_gen_meta(my_ukb_data)
my_gen_pcs &lt;- ukb_gen_pcs(my_ukb_data)
</code></pre></div></div>

<p><br /></p>

<p>A list of IDs for recommended exclusions and heterozygosity outliers (+/- 3*SD) can be retrieved</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_gen_excl(my_ukb_data)
ukb_gen_het(my_ukb_data)
</code></pre></div></div>

<p>For a <code class="highlighter-rouge">data.frame</code> of raw heterozygosity data</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_gen_het(my_ukb_data, all.het = TRUE)
</code></pre></div></div>

<p><br /></p>

<p><code class="highlighter-rouge">ukb_gen_rel</code> returns a <code class="highlighter-rouge">data.frame</code> with <code class="highlighter-rouge">id</code>, <code class="highlighter-rouge">pair</code> (a numeric identifier for related pairs), and <code class="highlighter-rouge">kinship</code> (kinship coefficient). For a count of related samples by degree of relatedness use <code class="highlighter-rouge">ukb_gen_rel_count</code>. Set the argument <code class="highlighter-rouge">plot = TRUE</code> to replicate the plot on page 15 of the <a href="http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/UKBiobank_genotyping_QC_documentation-web.pdf">UKB genotyping and quality control documentation</a>, for your subgroup of interest.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_gen_rel &lt;- ukb_gen_rel(my_ukb_data)

# To get a count and plot of degree of relatedness
ukb_gen_rel_count(my_gen_rel, plot = TRUE)
</code></pre></div></div>

<p><br /></p>

<h2 id="read-and-write">Read and write</h2>

<p>ukbtools includes functions to write phenotype and covariate files for <a href="https://jmarchini.org/bgenie/">BGENIE</a> and <a href="https://www.cog-genomics.org/plink2">PLINK</a>. <strong>BGENIE</strong> phenotype and covariate files are space-delimited, include column names, and have missing values coded as -999. They must also be in .sample file order. <code class="highlighter-rouge">ukb_gen_write_bgenie</code> sorts input data to match .sample file id order and writes the data to disk.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Read .sample file supplied with bulk genetic data
my_sample_file &lt;- ukb_gen_read_sample("path/to/sample_file")

# Write a BGENIE format phenotype or covariate file
ukb_gen_write_bgenie(
my_ukb_data, 
path = "path/to/bgenie_input_file", 
ukb.sample = my_sample_file, 
ukb.variables = c("variable1", "variable2", "variable3")
)
</code></pre></div></div>

<p><br /></p>

<p><strong>Note.</strong> The <a href="https://jmarchini.org/bgenie-usage/">BGENIE usage page</a> uses the example files example.pheno and example.cov - it is not clear whether the suffixes are obligatory. Use them to be safe.</p>

<p><br /></p>

<p><strong>PLINK</strong> phenotype and covariate files are either space- or tab-delimited, column names are optional, first two columns must contain family ID and individual ID respectively, and missing values are “normally expected to be encoded as -9” but also “nonnumeric values such as ‘NA’ … (are) treated as missing”. <code class="highlighter-rouge">ukb_gen_write_plink</code> writes a space-delimited file with column names, UKB ID is automatically written to column 1 and 2 and labelled FID IID, and missing values are coded as <code class="highlighter-rouge">NA</code>. The missing value to be used can be changed with the <code class="highlighter-rouge">na.strings</code> argument. See <a href="https://www.cog-genomics.org/plink/1.9/input#pheno">PLINK standard data input</a> for further details.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Write a PLINK format phenotype or covariate file

ukb_gen_write_plink(
my_ukb_data, 
path = "path/to/plink_input_file", 
ukb.variables = c("variable1", "variable2", "variable3")
)
</code></pre></div></div>

<p><strong>PLINK</strong> does not require that individuals in phenotype and covariate files are in any particular order, but you may want to reconcile the individuals you include in your analysis with those in the fam file (from your hard-called data). Read the fam file into R with <code class="highlighter-rouge">ukb_gen_read_fam</code></p>

<p><br /></p>

<h3 id="exclusions">Exclusions</h3>

<p><strong>Note.</strong> The exclusions referred to in this section are the combined UKB recommended exclusions, Affymetrix quality control for samples, bileve genotype quality control, heterozygosity outliers (+/- 3*SD), genetic ethnicity outliers (based on Genetic Ethnic Grouping, field code 22006), and one random member of each related pair (including Duplicates/MZ twins, 1st-, 2nd-, and 3rd-degree relatives).</p>

<p><br /></p>

<p><strong>BGENIE</strong> does not have an option to read exclusions. You can replace data values in a phenotype with <code class="highlighter-rouge">NA</code> where the individual is to-be-excluded based on genetic metadata considerations. Writing the updated variable to your phenotype file (with the supplied write functions), effectively excludes the individuals from any analysis.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_ukb_data$height_excl_na &lt;- ukb_gen_excl_to_na(my_ukb_data, x = "height")
</code></pre></div></div>
<p><br /></p>

<p><strong>PLINK</strong> <code class="highlighter-rouge">--remove</code> takes a space- or tab-delimited file with family IDs in the first column and individual IDs in the second column, without column names. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv">PLINK input filtering</a> for further details. (The missing value approach described above also works for PLINK.)</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ukb_gen_write_plink_excl("path/to/plink_input_file")
</code></pre></div></div>


      <footer class="site-footer">
        
          <span class="site-footer-owner"><a href="http://github.com/kenhanscombe/ukbtools">ukbtools</a> is maintained by <a href="http://github.com/kenhanscombe">kenhanscombe</a>.</span>
        
        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a>.</span>
      </footer>
    </section>

    
  </body>
</html>
