---
title: "Explore UK Biobank data"
author: "Ken B. Hanscombe"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<br>

After downloading and decrypting your UK Biobank (UKB) data with the supplied [UKB programs](http://biobank.ctsu.ox.ac.uk/crystal/docs/UsingUKBData.pdf), you have multiple files that need to be brought together to give you a tidy dataset to explore. The data file has column names that are the field codes from the UKB data showcase. `ukbtools` provides tools to tidy up the naming of the variables and perform exploratory analysis. 

<br>
<br>


## 1. Preparing a UKB fileset

Download^ยง^ then decrypt your data and create a UKB fileset (.tab, .r, .html):

```{bash, eval = FALSE}
ukb_unpack ukbxxxx.enc key
ukb_conv ukbxxxx.enc_ukb r
ukb_conv ukbxxxx.enc_ukb docs

```


`ukb_unpack` decrypts your downloaded `ukbxxxx.enc` file, outputting a `ukbxxxx.enc_ukb` file. `ukb_conv` with the `r` flag converts the decrypted data to a tab-delimited file `ukbxxxx.tab` and an R script `ukbxxxx.r` that reads the tab file. The `docs` flag creates an html file containing a field-code-to-description table (among others).

<br>

^ยง^ Full details of the data download and decrypt process are given in the [Using UK Biobank Data](http://biobank.ctsu.ox.ac.uk/crystal/docs/UsingUKBData.pdf) documentation .

<br>
<br>


## 2. Making a dataset

The function `ukb_df()` takes two arguments, the stem of your fileset and the path, and returns a dataframe with usable column names. 


```{r, eval = FALSE}

my_ukb_data <- ukb_df("ukbxxxx")

```


You can also specify the path to your fileset if it is not in the current directory. For example, if your fileset is in a subdirectory of the working directory called `data`


```{r, eval = FALSE}

my_ukb_data <- ukb_df("ukbxxxx", path = "/full/path/to/my/data")

```


<br>

__Note:__ You can move the three files in your fileset after creating them with `ukb_conv`, but they should be kept together. `ukb_df()` automatically updates the read call in the R source file to point to the correct directory (the current directly by default, or the directory specified by `path`).

<br>
<br>

### 2.1 Multiple downloads

If you have multiple UKB downloads, tidy then merge them.

```{r eval = FALSE}

ukbxxxx_data <- ukb_df("ukbxxxx")
ukbyyyy_data <- ukb_df("ukbyyyy")
ukbzzzz_data <- ukb_df("ukbzzzz")

my_ukb_data <- plyr::join_all(
  list(
    ukbxxxx_data,
    ukbyyyy_data,
    ukbzzzz_data),
  by = "eid",
  type = "full")

```

<br>

__Note:__ You will have at least one variable duplicated in the different downloads, the unique individual identifier `eid`. Counting the repeated variables (and the number of repeats of each) can give you a sense of the expected number of variables in your merged dataset.

```{r, eval = FALSE}

ukb_names <- c(
  names(ukbxxxx_data),
  names(ukbyyyy_data),
  names(ukbzzzz_data)
)
count_ukb_names <- data.frame(table(ukb_names))
dim(count_ukb_names[count_names$Freq > 1, ])

```


<br>
<br>


## 3. Retreiving ICD diagnoses


All ICD related functions begin `icd_`. Type `icd_` tab to see the family of functions. The full ICD "code-meaning" tables are available as datasets (`icd9codes`, `icd10codes`). ICD chapter-level tables describing disease blocks are also available for query as datasets (`icd9chapters`, `icd10chapters`)


<br>

To retrieve the full diagnosis of an individual

```{r, eval = FALSE}

icd_diagnosis(my_ukb_data, id = "0000000", icd.version = 10)

```

<br>
<br>

To retrieve the "meaning" of an ICD code use `icd_code`. (You can also look up multiple codes by concatenating them.)

```{r, eval = FALSE}

icd_code(icd.code = "H282", icd.version = 10)

```


<br>
<br>

To see the disease "blocks" that correspond to "chapters" in ICD-9 or ICD-10, `icd_chapter` takes the relevant ICD revision (or version) number

```{r, eval = FALSE}

icd_chapter(icd.version = 10)

```


<br>
<br>


## 4. Exploring a UKB sample subset

As an exploratory step you might want to look at the demographics of a particular subset of the UKB sample relative to a reference sample. For example, the subsample with data on your variable of interest compared to those without data (i.e. `NA`). `ukb_context` takes either a comparison variable (`comparison.var`) or a logical vector (`sample.ref`) to define a reference and comparison sample.

```{r, eval = FALSE}

ukb_context(
  my_ukb_data,
  age.var = "year_of_birth_0_0",
  sample.ref = sample(c(T,F), nrow(my_ukb_data), replace = TRUE)
  )

```


<br>
<br>


## 5. Genetic metadata

If you are doing any downstream genetic analyses, you will need the genetic metadata (which should be in you phenotype dataset). Detailed information is available on UKB [genotyping and quality control](http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/UKBiobank_genotyping_QC_documentation-web.pdf) and [imputation and association](http://www.ukbiobank.ac.uk/wp-content/uploads/2014/04/imputation_documentation_May2015.pdf).

All genetic metadata related functions begin `ukb_gen_`. Typing `ukb_gen_` tab opens a dropdown menu for this group of functions. 

<br>
<br>

You can collect the genetic metadata (including recommended exclusions, genetic sex, genetic ethnicity, chip, etc.), and principal components with

```{r, eval = FALSE}

my_gen_meta <- ukb_gen_meta(my_ukb_data)
my_gen_pcs <- ukb_gen_pcs(my_ukb_data)

```

<br>
<br>

A list of IDs for recommended exclusions and heterozygosity outliers (+/- 3*SD) can be retreived

```{r, eval = FALSE}

ukb_gen_excl(my_ukb_data)
ukb_gen_het(my_ukb_data)

```

For a dataframe of raw heterozygosity data

```{r, eval = FALSE}

ukb_gen_het(my_ukb_data, all.het = TRUE)

```


<br>
<br>

`ukb_gen_rel` returns a dataframe with `id`, `pair` (a numeric identifier for related pairs), and `kinship` (kinship coefficient)

```{r, eval = FALSE}

my_gen_rel <- ukb_gen_rel(my_ukb_data)

```


<br>
<br>


## 6. Read and write 

`ukbtools` includes functions to write phenotype and covariate files for [BGENIE](https://jmarchini.org/bgenie/) and [PLINK](https://www.cog-genomics.org/plink2). __BGENIE__ phenotype and covariate files are space-delimited, include column names, and have missing values coded as -999. They must also be in .sample file order. `ukb_gen_write_bgenie` sorts input data to match .sample file id order and writes the data to disk.


```{r, eval = FALSE}

# Read .sample file supplied with bulk genetic data
my_sample_file <- ukb_gen_read_sample("path/to/sample_file")

# Write a BGENIE format phenotype or covariate file
ukb_gen_write_bgenie(
  my_ukb_data, 
  path = "path/to/bgenie_input_file", 
  ukb.sample = my_sample_file, 
  ukb.variables = c("variable1", "variable2", "variable3")
)

```

<br>

__Note.__ The [BGENIE usage page](https://jmarchini.org/bgenie-usage/) uses the example files example.pheno and example.cov - it is not clear whether the suffixes are obligatory. Use them to be safe.

<br> 
<br>
 
__PLINK__ phenotype and covariate files are either space- or tab-delimited, column names are optional, first two columns must contain family ID and individual ID respectively, and missing values are "normally expected to be encoded as -9" but also "nonnumeric values such as 'NA' ... (are) treated as missing". `ukb_gen_write_plink` writes a space-delimited file with column names, UKB ID is automatically written to column 1 and 2 and labelled FID IID, and missing values are coded as `NA`. The missing value to be used can be changed with the `na.strings` argument. See [PLINK standard data input](https://www.cog-genomics.org/plink/1.9/input#pheno) for further details.


```{r, eval = FALSE}

# Write a PLINK format phenotype or covariate file
ukb_gen_write_plink(
  my_ukb_data, 
  path = "path/to/plink_input_file", 
  ukb.variables = c("variable1", "variable2", "variable3")
)

```

__PLINK__ does not require that individuals in phenotype and covariate files are in any particular order, but you may want to reconcile the individuals you include in your analysis with those in the fam file (from your hard-called data). Read the fam file into R with `ukb_gen_read_fam`


<br>
<br>

### 6.1 Exclusions

__Note.__ The exclusions referred to in this section are the combined UKB recommended exclusions, Affymetrix quality control for samples, bileve genotype quality control, heterozygosity outliers (+/- 3*SD), genetic ethnicity outliers (based on Genetic Ethnic Grouping, field code 22006), and one random member of each related pair (including Duplicates/MZ twins, 1st-, 2nd-, and 3rd-degree relatives).

<br>

__BGENIE__ does not have an option to read exclusions. You can replace data values in a phenotype with `NA` where the individual is to-be-excluded based on genetic metadata considerations. Writing the updated variable to your phenotype file (with the supplied write functions), effectively excludes the individuals from any analysis.


```{r, eval = FALSE}

my_ukb_data$height_excl_na <- ukb_gen_excl_to_na(my_ukb_data, x = "height")

```

<br>

__PLINK__ `--remove` takes a space- or tab-delimited file with family IDs in the first column and individual IDs in the second column, without column names. See [PLINK input filtering](https://www.cog-genomics.org/plink/1.9/filter#indiv) for further details. (The missing value approach described above also works for PLINK.)

```{r, eval = FALSE}

ukb_gen_write_plink_excl()

```
